### PR TITLE
Add InStruct() to Reader and Writer API

### PR DESCRIPTION
**Pull request by Armanbqt**
InStruct() needs to be in the API so IonHash can use it.

### CODE REVIEW

**Comment by Armanbqt** [comment]
This is to address #71 
Please let me know if there are any issues with changes of the API.

File: cmd/ion-go/eventwriter.go [hunk 1]
```diff
@@ -194,11 +195,13 @@ func (e *eventwriter) BeginStruct() error {
 		return err
 	}
 	e.depth++
+	e.inStruct = true
 	return nil
 }
 
 func (e *eventwriter) EndStruct() error {
 	e.depth--
+	e.inStruct = false
```
Comment (fernomac) [inline cmd/ion-go/eventwriter.go:204]: `{structs:{can:"end"},inside:{other:"structs"}}` :(

File: ion/textreader_test.go [hunk 2]
```diff
@@ -423,6 +423,38 @@ func TestTrsToString(t *testing.T) {
 	}
 }
 
+func TestInStruct(t *testing.T) {
+	r := NewReaderStr("[ { a:() } ]")
+
+	r.Next()
+	r.StepIn() // In the list, before the struct
+	if r.InStruct() {
+		t.FailNow()
```
Comment (fernomac) [inline ion/textreader_test.go:432]: `t.Fatal("InStruct returned true before we were in a struct")`?

File: ion/reader.go [hunk 3]
```diff
@@ -13,7 +13,7 @@ import (
 // A Reader reads a stream of Ion values.
 //
 // The Reader has a logical position within the stream of values, influencing the
-// values returnedd from its methods. Initially, the Reader is positioned before the
+// values returned from its methods. Initially, the Reader is positioned before the
```
Comment (fernomac) [inline ion/reader.go:16]: Thanks! :)

File: cmd/ion-go/eventwriter.go (hunk 1 above)
Comment (Armanbqt) [reply cmd/ion-go/eventwriter.go:204]: Good catch. I'll fix it.

File: ion/textreader_test.go (hunk 2 above)
Comment (Armanbqt) [reply ion/textreader_test.go]: Done.

File: ion/reader.go [hunk 4]
```diff
@@ -148,6 +148,9 @@ type Reader interface {
 	// ByteValue returns the current value as a byte slice (if that makes sense). It returns
 	// an error if the current value is not an Ion clob or an Ion blob.
 	ByteValue() ([]byte, error)
+
+	// InStruct indicates if the reader is currently positioned in a struct.
+	InStruct() bool
```
Comment (zslayton) [inline ion/reader.go:153]: Matter of taste: I'd prefer the name `IsInStruct` to make it obvious that this returns a `bool`.

File: cmd/ion-go/eventwriter.go [hunk 5]
```diff
@@ -15,6 +15,7 @@ type eventwriter struct {
 	depth       int
 	fieldname   string
 	annotations []string
+	inStruct    map[int]bool
```
Comment (zslayton) [inline cmd/ion-go/eventwriter.go:18]: Could we re-use the `ctxstack` concept that's leveraged by the reader? That would avoid heap allocations.

Comment (Armanbqt) [reply cmd/ion-go/eventwriter.go:18]: ```ctxstack``` is internal right now and can be only used within ```ion``` directory.
Do we want to make ```ctxstack``` public?

Comment (Armanbqt) [reply cmd/ion-go/eventwriter.go:18]: We can move ```ctx.go``` to 'internal' directory in order to be able use that in cmd package. But then we will have cycle imports (as ctx is using ion.Type).
A similar case is under discussion/investigation in an [ion-hash-go pull request](https://github.com/amzn/ion-hash-go/pull/31).

File: ion/reader.go (hunk 4 above)
Comment (Armanbqt) [reply ion/reader.go]: Renamed to ```IsInStruct```

File: cmd/ion-go/eventwriter.go (hunk 5 above)
Comment (Armanbqt) [reply cmd/ion-go/eventwriter.go:18]: To unlock Ion-hash, can we merge this pr and revisit this concern in #91 ?

**Pull request merged**